### PR TITLE
Full Python 3 support

### PR DESCRIPTION
**Pull request by drothlis**
Finally, the culmination of the Python 3 porting. With this PR it should work with both Python 2 & Python 3 (at least anything that's covered by the self-tests).

### TODO:

- [x] Merge #594 & rebase.
- [x] Get the integration tests passing.
- [x] Add integration tests with real lircd.

### Out of scope:

- Packaging.
- Figure out how to run `test_that_video_index_is_written_on_eos` on Python 3.

### CODE REVIEW

**Comment by drothlis** [comment]
@wmanley last one (hopefully) for you to review.

File: _stbt/core.py [hunk 1]
```diff
@@ -520,13 +520,13 @@ def test_that_the_on_screen_id_is_shown_after_booting():
...
               "re-raise it as PreconditionError.\nOriginal exception was:\n%s"
-              % (type(e).__name__, traceback.format_exc(e)))
-        exc = PreconditionError(message, e)
-        if hasattr(e, 'screenshot'):
-            exc.screenshot = e.screenshot  # pylint: disable=attribute-defined-outside-init,no-member
+              % (type(original).__name__, traceback.format_exc(original)))
+        exc = PreconditionError(message, original)
+        if hasattr(original, 'screenshot'):
+            exc.screenshot = original.screenshot  # pylint:disable=attribute-defined-outside-init,no-member
         raise exc
```
[the remark this answers is not in the document]
Comment (drothlis) [reply _stbt/core.py]: It's tricky if you want to keep compatibility with Python 2. The syntax isn't valid in Python 2 so you can't even put it in an if/else block. You'd have to put it in an entirely separate module that is only imported on Python 3.

The way [future.utils.raise_from][1] works, is it constructs a string from the reprs of the 2 exceptions and then `exec`s it. As you can imagine it doesn't work very well, and it didn't work here (I tried).

[1]: https://python-future.org/compatible_idioms.html#raising-exceptions

**Comment by drothlis** [comment]
@wmanley thanks for the review!

File: tests/test-match.sh [hunk 2]
```diff
@@ -211,8 +211,8 @@ test_match_nonexistent_template() {
 	EOF
     ! stbt run -vv test.py \
         || fail "Trying to match an non-existant template should throw"
-    cat log | grep -q "FAIL: test.py: IOError: No such file: idontexist.png" \
-        || fail "Didn't see 'IOError: No such file'"
+    cat log | grep -Eq "FAIL: test.py: (IO|OS)Error: No such file: idontexist.png" \
```
[the remark this answers is not in the document]
Comment (wmanley) [reply tests/test-match.sh]: To expand upon this:

```
$ git grep 'raise .*IOError'
_stbt/core.py:        raise IOError("No such file: %s" % filename)
_stbt/core.py:        raise IOError("Failed to load image: %s" % absolute_filename)
_stbt/imgutils.py:            raise IOError("No such file: %s" % relative_filename)
_stbt/imgutils.py:            raise IOError("Failed to load image: %s" %
stbt_virtual_stb.py:            raise IOError("x keymap file %r does not exist" % x_keymap)
tests/validate-ocr.py:        raise IOError('No such file or directory "%s"' % imgname)
```

[the remark this answers is not in the document]
Comment (drothlis) [reply tests/test-match.sh]: But we'd have to define aliases for `FileNotFoundError` etc on Python2.

[the remark this answers is not in the document]
Comment (wmanley) [reply tests/test-match.sh]: So python-future doesn't do this for us? That's a shame.  They are pretty much all `FileNotFoundError` so we could define a function:

```
def raise_file_not_found(message, filename):
    if python2:
        IOError(errno.ENOENT, message, filename)
    else:
        raise FileNotFoundError(errno.ENOENT, message, filename)
```

Or we could even define our own `FileNotFoundError` as you suggested.

There's overlap here with #447 and #529

**Pull request merged**